[scripts-audit] add guidelines for cmake (#18319)
microsoft/vcpkg: docs/maintainers/cmake-guidelines.md

  except those which are intended to be passed as multiple arguments.
  - Example:
  ```cmake
  set(working_directory)
  if(DEFINED arg_WORKING_DIRECTORY)
    set(working_directory "WORKING_DIRECTORY" "${arg_WORKING_DIRECTORY}")
  endif()
  # calls do_the_thing() if NOT DEFINED arg_WORKING_DIRECTORY,
  # else calls do_the_thing(WORKING_DIRECTORY "${arg_WORKING_DIRECTORY}")
  do_the_thing(${working_directory})
  ```
- There are no "pointer" parameters
  (where a user passes a variable name rather than the contents)
  except for out parameters.
- Undefined names are not referenced.
- Out parameters are only set in `PARENT_SCOPE`.
- `CACHE` variables are not used.
  - Exception: internal global variables to avoid duplicating work.
- `include()`s are only allowed in `ports.cmake` or `vcpkg-port-config.cmake`.
- `foreach(RANGE)`'s arguments _must always be_ natural numbers,
  and `<start>` _must always be_ less than or equal to `<stop>`.
  - This must be checked if necessary.
- All port-based scripts must use `include_guard(GLOBAL)`
  to avoid being included multiple times.
- `set(VAR )` should not be used. Use `unset(VAR)` to unset a variable,

>   and `set(VAR "")` to set it to an empty string.

Reviewer: personally I think `set(VAR [[]])` would be cleaner but if `set(VAR "")` works for you I am fine with it,

### CMake Versions to Require

- All CMake scripts, except for `vcpkg.cmake`,
  may assume the version of CMake that is present in the
  `cmake_minimum_required` of `ports.cmake`.
  - This `cmake_minimum_required` should be bumped every time a new version
    of CMake is added to `vcpkgTools.xml`, as should the
    `cmake_minimum_required` in all of the helper `CMakeLists.txt` files.
- `vcpkg.cmake` must assume a version of CMake back to 3.1 in general
  - Specific functions and options may assume a greater CMake version;
    if they do, make sure to comment that function or option
    with the required CMake version.

### Changing Existing Functions

- Never remove arguments in non-internal functions;
  if they should no longer do anything, just take them as normal and warn on use.
- Never add a new mandatory argument.

### Naming Variables

- `cmake_parse_arguments`: set prefix to `"arg"`
- local variables are named `snake_case`
- Internal global variable names are named `Z_VCPKG_`.

> - External experimental global variable names are named `X_VCPKG_`.

Reviewer 2: ```suggestion
- External experimental global variable names are prefixed with `X_VCPKG_`.

```

> - Internal functions are named `z_vcpkg_*`

Reviewer 2: ```suggestion
- Internal functions are prefixed with `z_vcpkg_`
```

  (where a user passes a variable name rather than the contents)
  except for out parameters.
- Variables are not assumed to be empty. If the variable is intended to be used locally, it must be explicitly initialized to empty with `set(FOO "")`.
- All variables expected to be inherited from the parent scope across an API boundary (i.e. not a file-local function) should be documented. Note that all variables mentioned in triplets.md are considered documented.
- Out parameters are only set in `PARENT_SCOPE`.
- `CACHE` variables are not used.
  - Exception: internal global variables to avoid duplicating work.
- `include()`s are only allowed in `ports.cmake` or `vcpkg-port-config.cmake`.
- `foreach(RANGE)`'s arguments _must always be_ natural numbers,
  and `<start>` _must always be_ less than or equal to `<stop>`.
  - This must be checked if necessary.
- All port-based scripts must use `include_guard(GLOBAL)`
  to avoid being included multiple times.
- `set(VAR )` should not be used. Use `unset(VAR)` to unset a variable,
  and `set(VAR "")` to set it to the empty value. _Note: this works for use as a list and as a string_

### CMake Versions to Require

> - All CMake scripts, except for `vcpkg.cmake`,

Reviewer 3: `vcpkg-cmake-wrapper.cmake` scripts run in the same context as `vcpkg.cmake` (aka project mode), so they fall under the same exception.

Reviewer 4: Ports can require a version of CMake that is later than what `vcpkg.cmake` cares about.

  may assume the version of CMake that is present in the
  `cmake_minimum_required` of `ports.cmake`.
  - This `cmake_minimum_required` should be bumped every time a new version
    of CMake is added to `vcpkgTools.xml`, as should the
    `cmake_minimum_required` in all of the helper `CMakeLists.txt` files.

> - `vcpkg.cmake` must assume a version of CMake back to 3.1 in general

Reviewer 3: For `vcpkg-cmake-wrapper.cmake` scripts, version 3.1 even excludes `if("value" IN_LIST FEATURES)`. This makes it more complicated to deal with port features or even with arguments to `find_package` (e.g. port tiff).

Reviewer: > `vcpkg.cmake` must assume a version of CMake back to 3.1 in general

Is sentence is nonsense and should be ignored. I think the correct phrasing would be that `vcpkg.cmake` must assume that the toplevel `CMakeLists.txt` might not have a `cmake_minimum_required` or a very low `cmake_minimum_required`. As such, setting appropriate cmake policies are a requirement. vcpkg assume that it is at least executed with a cmake binary of <some cmake version|preferly the cmake version used by vcpkg>. 
If there is somebody really arguing for a cmake binary of 3.1 I would like to have him/her invent `vcpkg.old.cmake` .....

Reviewer 4: No, we really do support CMake version back to CMake 3.1 in vcpkg.cmake. There are extra features that require more recent versions, but we do support back to CMake 3.1 (or there's a bug.)

Reviewer 3: What is the reason to keep it at 3.1?
What is the threshold for moving it to another version?
As I already wrote:
> For vcpkg-cmake-wrapper.cmake scripts, version 3.1 even excludes if("value" IN_LIST FEATURES).

which is relevant for dealing with transitive dependencies.

Reviewer 4: we do not consider ports as having to support any specific version of CMake; they can support whatever they want. All we need is that, if a port supports CMake 3.1, and the user uses 3.1 for their build, it works.

# CMake Guidelines

We expect that all CMake scripts that are either:

- In the `scripts/` directory, or
- In a `vcpkg-*` port

should follow the guidelines laid out in this document.
Existing scripts may not follow these guidelines yet;
it is expected that we will continue to update old scripts to fall in line with these guidelines.

These guidelines are intended to create stability in our scripts.
We hope that they will make both forwards and backwards compatibility easier.

## The Guidelines

- We always use `cmake_parse_arguments` rather than function parameters, or referring to `${ARG<N>}`.
  - This doesn't need to be followed for "script-local helper functions"
  - Exception: exclusively positional parameters, like out variables.
    - In this case, positional parameters should be put in the function declaration
      (rather than using `${ARG<N>}`), and should be named according to local rules
      (i.e. `snake_case`).
    - Exception: positional parameters that are optional should be given a name via
      `set(argument_name "${ARG<N>}") after checking `${ARGC}`.

> - There are no unparsed or unused arguments. Always check for `ARGN` or `arg_UNPARSED_ARGUMENTS`,

Reviewer: This in generally leads to a lot of duplication of parameters in chained function calls

e.g. f1 calling f2 calling f3; f1 basically needs to know all possible options for f2 and f3.  
Having `arg_UNPARSED_ARGUMENTS` also requires to have an option or a one value parameter in the function, The last argument is a multi value parameter there will never be `arg_UNPARSED_ARGUMENTS` because it will be consumed by the multi value parameter.

Reviewer: So basically function calls need to be `functioncall(<multi_var_args> <one_value_args> <options>)`

Reviewer 4: This doesn't make sense to me. `functioncall(... MULTIVALUE asdf asdf asdf)` has no unparsed arguments.

Reviewer: > functioncall(... MULTIVALUE asdf asdf asdf) has no unparsed arguments.

That is the problem. Let it simply be `functioncall(... MULTIVALUE asdf asdf OUTDATED_OPTION)`. Now you logically would have a `arg_UNPARSED_ARGUMENTS` but cannot detect it. 
Ah I see I meant `functioncall(<options> <one_value_args> <multi_var_args>)` should be the rule

  and either `FATAL_ERROR`, or `WARN` if necessary for backwards compatibility.
- All `cmake_parse_arguments` use `PARSE_ARGV` for resistance to embedded semicolons.
- All `foreach` loops use `IN LISTS` for resistance to embedded semicolons.
- The variables `${ARGV}` and `${ARGN}` are unreferenced, except in helpful messages to the user.
  - (i.e., `message(FATAL_ERROR "blah was passed extra arguments: ${ARGN}")`)
- We always use functions, not macros or top level code.

>   - Exception: `vcpkg.cmake`'s `find_package`

Reviewer: ```suggestion
  - Exception: `vcpkg.cmake`'s `find_package`. The reason for this exceptions is that `find_package` is basically a special `include` statement. Parent scoping all newly set variables within the scope of the override would severely litter users `--trace/--trace-expand` output making it unusable. 

```

- Scripts in the scripts tree should not be expected to need changes as part of normal operation. (For example, `vcpkg_acquire_msys` has hard coded specific packages and versions thereof used which we believe is unacceptable)
- All non-splat variable expansions are in quotes "".
- There are no "pointer" parameters (where a user passes a variable name rather than the contents) except for out parameters.
- Undefined names are not referenced.
- Out parameters only set `PARENT_SCOPE`.
- `CACHE` variables are not used, except for internal global variables which are used to avoid duplicating work.
- `include()`s are removed and fixes to `port.cmake` et al. are made as necessary to avoid this.
- `foreach(RANGE)`'s arguments _must always be_ natural numbers, and `<start>` _must always be_ less than or equal to `<stop>`.
  - This should be checked.

> 

Reviewer: I am missing usage of https://cmake.org/cmake/help/git-stage/command/include_guard.html
and also a note weather or not scripts should include all other required scripts for operation (which is currently not the case for some reason).

Reviewer 4: TIL about `include_guard`

Reviewer 4: From latest:
 - `include()`s are only allowed in `ports.cmake` or `vcpkg-port-config.cmake`.

### Naming Variables

> - `cmake_parse_arguments`: set prefix to `"arg"`

Reviewer: Sorry but the prefix should be unique for each function. Otherwise you have a disaster with `arg_OPTIONS` in f1 calling f2 with possibly the same OPTIONS parameter. 
I would suggest using `CMAKE_CURRENT_FUNCTION` instead and using it as `set(z_arg ${CMAKE_CURRENT_FUNCTION})`
`cmake_parse_arguments(PARSE_ARGV <N> "${z_arg}" "OPTIONS" `and using `${${z_arg}_OPTIONS}`

Reviewer: i would also suggest to use `vcpkg_parse_arguments` here as a macro which automatically does the setting of `z_arg`

Reviewer 4: No this doesn't make any sense.

Reviewer: > No this doesn't make any sense. This is not how CMake's function scoping works

No that is wrong! But you could say that is not how `cmake_parse_arguments` works
```
include(CMakePrintHelpers)
function(foo)
    cmake_parse_arguments(PARSE_ARGV 0 arg "OPTION" "ONE_VALUE" "MULTI_VALUE")
    bar()
endfunction()

function(bar)
    cmake_print_variables(arg_OPTION arg_ONE_VALUE arg_MULTI_VALUE)
    cmake_parse_arguments(PARSE_ARGV 0 arg "OPTION" "ONE_VALUE" "MULTI_VALUE")
    cmake_print_variables(arg_OPTION arg_ONE_VALUE arg_MULTI_VALUE)
endfunction()

foo(OPTION "TRUE" ONE_VALUE "Hello" MULTI_VALUE "World" "!")
```
Output
```
$ cmake.exe -P function_scope.cmake
-- arg_OPTION="TRUE" ; arg_ONE_VALUE="Hello" ; arg_MULTI_VALUE="World;!"
-- arg_OPTION="FALSE" ; arg_ONE_VALUE="" ; arg_MULTI_VALUE=""

```

Reviewer 2: Just to wrap this back up: the case where this scoped overriding would be a problem is if the caller expected `arg_OPTION` to _dereference_ correctly within the child function's context. However, this is explicitly disallowed above due to

> - There are no "pointer" parameters
  (where a user passes a variable name rather than the contents)
  except for out parameters.

  - This must be checked if necessary.
- All port-based scripts must use `include_guard(GLOBAL)`
  to avoid being included multiple times.

> - `set(VAR )` should not be used. Use `set(VAR)` to unset a variable.

Reviewer 4: I dunno, it seems fine to me?

- All port-based scripts must use `include_guard(GLOBAL)`
  to avoid being included multiple times.

### CMake Versions to Require

- All CMake scripts, except for `vcpkg.cmake`,
  may assume the version of CMake that is present in the

>   `cmake_minimum_required` of `ports.cmake`.

Reviewer: Are scripts allowed to bump the version up? E.g. it would be helpful to have access to string(JSON) which is 3.19 to parse the version of a port from the manifest in the portfile.

Reviewer 4: We'll bump it up as we update; that's a good point though.
